### DISCUSSION TITLE
Bit-packed GF2

### DISCUSSION THREAD

**Discussion by amirebrahimi** [mhostetter/galois#502] [Ideas]
Hi Matt,

We've been using GF2 for a project and it's been quite useful, but have since needed to move to bit-packing for performance reasons. Is this something you've ever thought about adding or would want to have included in the package (only for GF2 specifically). We have a few routes we can move forward with, such as rolling our own package or possibly contributing some work back to Galois, so I figured I'd ask.

Cheers,
Amir

**Comment by mhostetter** [comment]
Hey Amir, thanks for reaching out. I did something similar with bit-packing GF(2) arrays in the past. What kinds of operations do you desire with a bit-packed array? Simple addition and multiplication? Or more complex things like matrix multiplication?

I can see a use-case here. There would probably need to be a separate class, perhaps `GF2BitPacked` or `GF2Packed`? Supporting addition and multiplication is easy, but some other operations are trickier and would require custom ufuncs. I think adding something like this could be useful to the library.

**Reply by amirebrahimi** [reply]
Yes, matrix multiplication would be needed. Occasionally we use other ufuncs like count_nonzero, hsplit, hstack, vstack, all, etc.

**Reply by amirebrahimi** [reply]
As far as a separate class, it would be ideal if `bitpacked=True` could be passed instead of creating a whole separate class and just have the internals handled differently in that case. If it must be a separate class, I'd assume we could inherit from `GF2`?

**Reply by mhostetter** [reply]
Every type of field is a unique `FieldArray` subclass. So `GF = galois.GF(7)` returns a new class that defines the arithmetic for that field. That way `x.view(GF)` will work as intended. The bit-packed GF(2) class will need to be separate from the conventional GF(2) class because the ufuncs will differ. We can, however, make the API convenient. So `GF = galois.GF(2, bitpacked=True)` is easy enough. But technically that would return a different class than `GF = galois.GF(2)`.

What types of dtypes do you desire support for? Just uint8 or all unsigned integers?

I'm trying to understand how the bit-packing would work with > 1D arrays. For instance, with matrix multiplication, can you give me an example of the dimensions of the arrays and the number of packed bits?

For instance, below is a 2x1 array of uint8s. Would you consider this a 2x8 conventional (non-packed) matrix?

```python
In [3]: x = np.array([[0b01001100], [0b11100010]]); x
Out[3]: 
array([[ 76],
       [226]])
```

How would you multiply a conventional 8x3 matrix to this one? It needs 8 rows, which is easy. But it also needs 3 columns. How would a bit-packed array with uint8s have 3 columns, since each element contains 8 values?

Maybe I'm missing something obvious...

**Reply by amirebrahimi** [reply]
I think `uint8` is sufficient. Possibly `bool`, too.
We'd follow the same convention as `np.packbits`, which pads the arrays to full bytes by adding 0s.
Unfortunately, for matrix multiplication we'd still incur a hit with unpacking bits.

**Reply by mhostetter** [reply]
I think that sounds good. Potentially we can intercept the call to `np.packbits()` such that `GF2` instances passed into `np.packbits()` are returned automatically as `GF2Packed` instances. The same could be done for `np.unpackbits()`.

If you all are interested in submitted a PR, I'm open to adding this to the library. I can provide some pointers on where I think functionality will need to be added.

**Reply by oscarhiggott** [reply]
I'd also find it useful if there was an option for GF2 to be bitpacked. Would row operations e.g. in `GF2.row_reduce` be vectorised, and if so would it also make sense to support bitpacking into 64-bit integers to improve performance?

**Reply by amirebrahimi** [reply]
@mhostetter, I've started work on this now. Here is a first pass:
https://github.com/mhostetter/galois/compare/main...amirebrahimi:galois:main

If anything feels off with the direction please direct me towards any classes / functionality I might not know about.

**Reply by amirebrahimi** [reply]
One thing I am wondering is how to restrict this support, so only GF2 can be allowed to have `bitpacked=True`.

**Reply by amirebrahimi** [reply]
After working on this a bit more I'm now playing a game of whack-a-mole with numpy functions. So, I think I'm going to go the route of restricting this class to only allow operations that benefit from being bit-packed (e.g. addition/element-wise multiplication and possibly matmul with different bit-packed axes). Then, there can be follow-up ticket(s) to expand scope to a more implicit bitpacked data structure that will expand and pack as necessary.

**Reply by mhostetter** [reply]
Thanks for your work on this. I feel there are a lot of details to discuss and think through. Would it be best to open a PR (target `release/0.4.x`) and we can comment on them iteratively there? You can mark as "draft", if you'd like.

**Reply by amirebrahimi** [reply]
Sure thing, please take a look at #583.

**Comment by amirebrahimi** [comment]
Sounds good. We haven't started on this work yet, so we are open to any pointers you wish to provide.
